### PR TITLE
Database migrations

### PR DESCRIPTION
**Pull request by The0mikkel**
***This update implements database migration***

## What is a migration?
A migration, is a piece of sql code, that is designed to move the database from one design to another.<br>
Ie. Adding new tables, changing existing tables or dropping tables.
This can also be used to seed the database.

For more information: 
- [What are database migrations? - Prisma](https://www.prisma.io/dataguide/types/relational/what-are-database-migrations)
- [What is Database Migration? - Aloowma](https://www.alooma.com/blog/what-is-database-migration)

## How does it work?
A new class - DBMigration - has been added, to handle all migrations.

The system is build around the "migrations" folder in `src\main\resources\dk\sdu\se_f22\sharedlibrary\db\migrations`, where each .sql file is handled as a migration.

Files are sorted alphabetically and run one by one.

Each migration is logged in the "migrations" table, and files that have already been run, will not be run again.<br>
A batch system also makes it possible, to see when a migration has been made.

## The idea with the migration .sql files
The idea behind the files is, that each update to the database, is kept in a new file.<br>
The files may be named as followed (naming is only used for sorting - Very important for the order the migrations should be done in):<br>
\<version>.\<minor_version>_\<descriptive name>.sql

If the file does not end with ".sql" it will not be categorized as a migration.

## Continues migration
The system is built to be able to be run either once when the individual developer needs to run the migration or can be implemented as a step when running the full system.

A command-line print tool is provided, that can be disabled, to print the status of the migration to the console. This uses custom colors so that developers can better sort through all the migration steps.

## Migration types
I have included two types of migration:
- **migration** - *.migrate()* - The normal migration, that simply updates the database.
- **Fresh migration** - *-migrateFresh()* - This will flush the database, on the "public" schema, and recreate the whole database, from scratch - This will delete data and allows for developers to test the system on a fresh database, without any previous data. 

***All information in this pull request should be available either in the readme or as JavaDoc comments in the new class.***

## Tl;dr
This makes it possible to update the database. The only thing you have to do, is write a new .sql file in `src\main\resources\dk\sdu\se_f22\sharedlibrary\db\migrations` that contains the sql code necessary to update the database to the new desired state. As well as run the migration code.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -44,3 +44,75 @@ javase --- sharedlibrary
...
+- **db_user**: Username of the user, that you want to access the database as
+- **db_password**: The password for the provided database user
+
+***The `config.properties.example` is an example file, to show how the `config.properties` file should look. The data in these files, will not be commited to git***
+
+### Migration
+
+The system, has a migration tool build in, that can be used to keep the database up to date.
+
+This migration system is location in `dk\sdu\se_f22\sharedlibrary\db\DBMigration.java`
```
Comment (mathiasgredal) [inline README.md:65]: Perhaps add a direct clickable link to the file

File: README.md [hunk 2]
```diff
@@ -44,3 +44,75 @@ javase --- sharedlibrary
...
+
+The migration system, has build-in command line printing, and can be disabled by sending a `false` argument with the constructor:
```
Comment (mathiasgredal) [inline README.md:67]: built-in

File: README.md [hunk 3]
```diff
@@ -44,3 +44,75 @@ javase --- sharedlibrary
...
+```java
+DBMigration databaseSeeder = new DBMigration(false);
+```
+
+
+#### What is a migration
+
+A migration, is a piece of sql code, that is designed to move the database from one design to another.<br>
+Ie. Adding new tables, changing existing tables or dropping tables.
```
Comment (mathiasgredal) [inline README.md:76]: Lowercase: i.e

File: README.md [hunk 4]
```diff
@@ -44,3 +44,75 @@ javase --- sharedlibrary
...
+
+The system allows for the database to be freshly migrated.<br>
+***THIS WILL CLEAR ALL OF THE DATA FROM THE DATABASE PUBLIC SCHEMA***<br>
+*Use with caution*
+```java
+DBMigration databaseSeeder = new DBMigration();
+databaseSeeder.migrateFresh();
+```
+
+It is not possible to rollback any migrations, but if a migration fails, the system will hald, and automaticly rollback to the last stable migration.
```
Comment (mathiasgredal) [inline README.md:101]: halt and automatically

File: README.md [hunk 5]
```diff
@@ -44,3 +44,75 @@ javase --- sharedlibrary
...
+
+The current system, is only one way migration, and no rollback function is available. **If a migration has been run, it cannot be rolledback.**
+
+#### How to write a migration
+
+The migrations, is located in the folder `src\main\resources\dk\sdu\se_f22\sharedlibrary\db\migrations`.
```
Comment (mathiasgredal) [inline README.md:107]: are

File: README.md [hunk 6]
```diff
@@ -44,3 +44,75 @@ javase --- sharedlibrary
...
+
+All files, that end in .sql will be runned, as if it was a migration.<br>
+The system automaticly keeps track of which files has been runned.<br>
```
Comment (mathiasgredal) [inline README.md:110]: automatically

File: README.md [hunk 7]
```diff
@@ -44,3 +44,75 @@ javase --- sharedlibrary
...
+**Please note, when changing the names of a migration file, the system will treat this as a new migration.**
+
+The files may be named as followed (naming is only used for sorting - Very important for the order the migrations should be done in):<br>
```
Comment (mathiasgredal) [inline README.md:113]: follows

File: README.md [hunk 8]
```diff
@@ -44,3 +44,75 @@ javase --- sharedlibrary
...
+\<version>.\<minor_version>_\<descriptive name>.sql
+
+Take a look int the existing migrations, to see examples, of how they are build.
```
Comment (mathiasgredal) [inline README.md:116]: Should it be “in” or “at”, can’t be both? Also built

File: src/main/java/dk/sdu/se_f22/sharedlibrary/db/DBMigration.java [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+
+        // creates a file object
+        File file = new File(migrationsPath);
+
+        // returns an array of all files, sorted alphabetically
+        String[] fileList = file.list();
+        Arrays.sort(fileList, new NaturalOrderComparator());
+
+        // Get known migrations
+        List migrations = this.getMigrations();
```
Comment (mathiasgredal) [inline src/main/java/dk/sdu/se_f22/sharedlibrary/db/DBMigration.java:75]: Can this be of type `List<String>` instead?

File: src/main/java/dk/sdu/se_f22/sharedlibrary/db/DBMigration.java [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+        this.batch++;
+
+        boolean migrationStatus;
+
+        for (String fileName : fileList) {
+            // Ensure the file is a sql file
+            String lowercaseFileName = fileName.toLowerCase(); 
+            if (!lowercaseFileName.endsWith(".sql")) {
+                continue;
+            }
```
Comment (mathiasgredal) [inline src/main/java/dk/sdu/se_f22/sharedlibrary/db/DBMigration.java:85]: Method extraction or something, idk

File: src/main/java/dk/sdu/se_f22/sharedlibrary/db/DBMigration.java [hunk 11]
```diff
@@ -0,0 +1,267 @@
...
+    }
+
+    /**
+     * migrate the database to a fresh (new) state - THIS WILL DELETE ALL DATA
+     */
+    public void migrateFresh() {
+        this.println("Flushing database", Color.YELLOW);
+
+        try {
+            PreparedStatement stmt;
```
Comment (mathiasgredal) [inline src/main/java/dk/sdu/se_f22/sharedlibrary/db/DBMigration.java:119]: Remember to close prepared statements

File: src/main/java/dk/sdu/se_f22/sharedlibrary/db/DBMigration.java [hunk 12]
```diff
@@ -0,0 +1,267 @@
...
+            // Drop schema to delete all tables
+            stmt = connection.prepareStatement("DROP SCHEMA public CASCADE;");
+            stmt.execute();
+
+            // Recreate schema
+            stmt = connection.prepareStatement("CREATE SCHEMA public;");
+            stmt.execute();
+        } catch (SQLException ex) {
+            ex.printStackTrace();
```
Comment (mathiasgredal) [inline src/main/java/dk/sdu/se_f22/sharedlibrary/db/DBMigration.java:128]: Why not use your fancy printing methods?

File: src/main/java/dk/sdu/se_f22/sharedlibrary/db/DBMigration.java [hunk 13]
```diff
@@ -0,0 +1,267 @@
...
+            connection.setAutoCommit(true);
+        } catch (SQLException e) {
+            e.printStackTrace();
+        }
+        return true;
+    }
+
+    private List<String> getMigrations() {
+        try {
+            PreparedStatement stmt = connection.prepareStatement("SELECT file_name, batch FROM migrations ORDER BY batch");
```
Comment (mathiasgredal) [inline src/main/java/dk/sdu/se_f22/sharedlibrary/db/DBMigration.java:229]: Again you should close prepared statements and resultsets.

File: src/main/java/dk/sdu/se_f22/sharedlibrary/db/DBMigration.java [hunk 14]
```diff
@@ -0,0 +1,267 @@
...
+            return returnValue;
+        } catch (SQLException ex) {
+            // ex.printStackTrace();
+            return new ArrayList<String>();
+        }
+    }
+
+    private boolean addMigration(String fileName) {
+        try {
+            PreparedStatement stmt = connection.prepareStatement("INSERT INTO migrations (file_name, batch) VALUES (?, ?)");
```
Comment (mathiasgredal) [inline src/main/java/dk/sdu/se_f22/sharedlibrary/db/DBMigration.java:247]: And here

**Review by mathiasgredal — COMMENTED** [review]
Minor nitpicks, looks good overall😀?

**Review by v-nemeth — COMMENTED** [review]
Det ser super fint ud, og er til at forstå, ved dog ikke helt hvad du har tænkt dig at bruge farverne i Color.java eller NaturalOrderComparator til

**Comment by The0mikkel** [comment]
> Det ser super fint ud, og er til at forstå, ved dog ikke helt hvad du har tænkt dig at bruge farverne i Color.java eller NaturalOrderComparator til

Thanks :)

#### Color.java
Color.java, is an enum, that holds color codes for colors, that can be used to color the text in a command prompt.<br>
This makes it possible, to make eye-catching printing.

This is used in the migration, to quickly inspect what the tool has done:
- Yellow: Running migration
- Green: Migration is done
- Green with underline: Database is up to date, and all migrations has been run successfully
- Red bold/bright: Error - Something went very wrong, and the migration process should have stopped, due to a failure
- Red: Most likely the stack trace - A less important error, or context for the error.

#### NaturalOrderComparator.java
A problem arises when sorting strings. It is easy to do, if you ONLY look at the characters, and their place in the alphabet, but what happens when you mix in numbers and other fun things?

Well, then "11 something" will come before "2 something". This is especially bad when we talk migrations, that start with a number. I don't want to run migration 10, before migration 2.

For more information on what natural sort is, see [https://en.wikipedia.org/wiki/Natural_sort_order](https://en.wikipedia.org/wiki/Natural_sort_order).

The implementation, which is located in NaturalOrderComparator.java, is an implementation of natural sort for objects, not just Strings actually, that I have taken from Stackoverflow - See author field on the class - because I didn't believe that I can write a better implementation myself, nor saw the time for it.

File: src/main/java/dk/sdu/se_f22/sharedlibrary/db/DBMigration.java [hunk 15]
```diff
@@ -0,0 +1,313 @@
...
+                if (currentLine.contains("--"))
+                    continue;
+
+                // Safeguards commented code in /* comment /n */ format
+                if (currentLine.contains("/*")) {
+                    // Safeguards /* */ in same line
+                    if (currentLine.contains("*/"))
+                        break;
+
+                    String commentedLine;
```
Comment (OJCJ) [inline src/main/java/dk/sdu/se_f22/sharedlibrary/db/DBMigration.java:175]: Variable never initialised.

**Review by OJCJ — APPROVED** [review]
Looks good 😁👍
